### ISSUE TITLE
Enshrine head state shuffling in the `shuffling_cache`

### ISSUE THREAD

**Issue by paulhauner** [sigp/lighthouse#4281]
## Description

During the mainnet instability on the 12th & 13th of May we saw some logs like this:

```
ERRO Failed to download attester duties      err: FailedToDownloadAttesters("Some endpoints failed, num_failed: 1 http://redacted:5052/ => RequestFailed(ServerMessage(ErrorMessage { code: 500, message: "UNHANDLED_ERROR: MaxCommitteePromises(2)", stacktraces: [] }))"), request_epoch: 200554, current_epoch: 200553, service: duties
```

This is a result of the DoS prevention mechanism that prevents us from concurrently loading and unbounded number of states from the DB to serve requests for shuffling (triggered via P2P or HTTP API): 

https://github.com/sigp/lighthouse/blob/693886b94176faa4cb450f024696cb69cda2fe58/beacon_node/beacon_chain/src/shuffling_cache.rs#L14-L23

This mechanism is likely the one that kept us stable during the network instability. However, we did notice that our [shuffling cache](https://github.com/sigp/lighthouse/blob/stable/beacon_node/beacon_chain/src/shuffling_cache.rs) doesn't enshrine the shuffling of the head block in the same way that our [`snapshot_cache` does](https://github.com/sigp/lighthouse/blob/693886b94176faa4cb450f024696cb69cda2fe58/beacon_node/beacon_chain/src/snapshot_cache.rs#L225-L229). By "enshrine" I mean that the cache should know what the attester shuffling key is for the head block (at the current wall-clock epoch) and it should not allow some other shuffling to replace it. This should prevent us from ever dropping attestations to the head block because we've reached `MaxCommitteePromises`.

## Steps to Resolve

From a high level, I think we need to:

1. Store the `AttestationShufflingId` for the head block and current wall-clock epoch on the `ShufflingCache`
    - It might actually turn out to be best to just store the shuffling decision root for the head block on the `ShufflingCache` and also provide the cache with access to the `SlotClock`, so it can figure out the current wall-clock epoch on its own accord.
1. Ensure that any insertions to the cache (e.g., `self.cache.put`) don't override the enshrined head values.

I think (1) is fairly straight-forward. The `shuffling_cache` already provides an example of how this is done.

(2) is a little more complex since the shuffling cache uses an `LruCache` which natively just replaces the oldest entry rather than a specific one. I see two ways to address this:

- Touch the enshrined head shuffling before any insert, so it's always the most recent and least-likely to be removed.
- Switch away from an LruCache to something that prefers to remove the entry with the lowest epoch (this is how the `snapshot_cache` does it).

At this moment, I prefer the latter since it seems a little smarter than just an LRU cache. However I'm not convinced since there are some complexities around tie-breaking when two entries have the same epoch. Perhaps preferring entries in the canonical chain is best? It's pretty late here so I'd have to come back to this to reason it out properly.

**Comment by michaelsproul** [comment]
It might also be good to make the concurrency limit configurable via the CLI. Then users can throw more CPU at it if the default is too low

**Comment by jimmygchen** [comment]
Hi, I'll look into this one!

**Comment by michaelsproul** [comment]
Fixed in #4296!
